Add Bevy 0.17 release notes draft (#2233)
bevyengine/bevy-website: content/news/2025-09-24-bevy-0.17/index.md

+++

Thanks to **X** contributors, **X** pull requests, community reviewers, and our [**generous donors**](/donate), we're happy to announce the **Bevy 0.17** release on [crates.io](https://crates.io/crates/bevy)!

For those who don't know, Bevy is a refreshingly simple data-driven game engine built in Rust. You can check out our [Quick Start Guide](/learn/quick-start) to try it today. It's free and open source forever! You can grab the full [source code](https://github.com/bevyengine/bevy) on GitHub. Check out [Bevy Assets](https://bevy.org/assets) for a collection of community-developed plugins, games, and learning resources.

To update an existing Bevy App or Plugin to **Bevy 0.17**, check out our [0.16 to 0.17 Migration Guide](/learn/migration-guides/0-16-to-0-17/).

Since our last release a few months ago we've added a _ton_ of new features, bug fixes, and quality of life tweaks, but here are some of the highlights:

- **X:** x
<!-- more -->

## Experimental Raytraced Lighting: Bevy Solari

{{ heading_metadata(authors=["@JMS55", "@SparkyPotato"] prs=[19058, 19620, 19790, 20020, 20113, 20156, 20213, 20242, 20259, 20406, 20457, 20580, 20596, 20622, 20658, 20659, 20980]) }}

> (TODO: Embed solari example screenshot here)

Reviewer: Use a video instead tbh. Especially showing camera motion.

> With the new `bevy_solari` crate, we're making our first steps towards realtime raytraced lighting.

Reviewer: I think this section should be more exciting, less theory.

### Try it out

While Bevy 0.17 adds the bevy_solari crate, it is not yet production ready.

However, feel free to run the solari example to check out the progress we've made. There are two different modes you can try out:

A non-realtime "reference" mode that uses pathtracing: `cargo run --release --example solari --features bevy_solari -- --pathtracer`.

A realtime mode that uses a combination of techniques, and currently supports only diffuse materials: `cargo run --release --example solari --features bevy_solari`.

Additionally, if you have a NVIDIA GPU, you can enable DLSS Ray Reconstruction with the realtime mode for a combination of denoising (Bevy Solari does not currently come with any alternate denoisers), lower rendering times, and anti-aliasing: `cargo run --release --example solari --features bevy_solari,dlss`.

### How it works

> TODO: Showcase the different aspects/steps of Solari (direct initial+temporal+spatial, indirect initial+temporal+spatial, world cache, DLSS RR)

Reviewer: TODO: link blog

[`GlobalTrigger`]: https://dev-docs.bevy.org/bevy/ecs/event/struct.GlobalTrigger.html
[`EntityEvent`]: https://dev-docs.bevy.org/bevy/ecs/event/trait.EntityEvent.html
[`ChildOf`]: https://dev-docs.bevy.org/bevy/ecs/hierarchy/struct.ChildOf.html
[`PropagateEntityTrigger`]: https://dev-docs.bevy.org/bevy/ecs/event/struct.PropagateEntityTrigger.html
[`Add`]: https://dev-docs.bevy.org/bevy/ecs/lifecycle/struct.Add.html
[`EntityComponentsTrigger`]: https://dev-docs.bevy.org/bevy/ecs/event/struct.EntityComponentsTrigger.html
[`AnimationPlayer`]: https://dev-docs.bevy.org/bevy/animation/struct.AnimationPlayer.html
[`AnimationEvent`]: https://dev-docs.bevy.org/bevy/animation/trait.AnimationEvent.html
[`AnimationEventTrigger`]: https://dev-docs.bevy.org/bevy/animation/struct.AnimationEventTrigger.html

## Experimental UI Widgets: Bevy Feathers

{{ heading_metadata(authors=["@viridia", "@Atlas16A", "@ickshonpe", "@amedoeyes"] prs=[19730, 19900, 19928, 20237, 20169, 20422, 20350, 20548, 20969]) }}

TODO: add screenshot of `feathers` in practice

To make it easier for Bevy engine developers and third-party tool creators to make comfortable, visually cohesive tooling,

> we're pleased to introduce "Feathers" - a comprehensive Bevy UI widget set. Feathers is meant to be Bevy's "developer tools" widget set, and it will be used to build the upcoming [Bevy Editor](https://bevy.org/news/bevys-fifth-birthday/#bevy-editor-design). It has a utilitarian look and feel, with an opinionated feature-set tailored for editors and graphical utilities. It builds on top of Bevy's new general-purpose "headless" widget set: `bevy_ui_widgets` (covered below). Feathers _can_ be used in games, but that is not its motivating use case.

Reviewer: I feel like we might want to move bevy_ui_widgets to below this.

- We will port Feathers to BSN (Bevy's [Next-Generation Scene/UI System](https://github.com/bevyengine/bevy/pull/20158/)) when that lands (targeted for **Bevy 0.18**).
- We are still discussing the best way to handle UI callbacks / events in Feathers. It includes a proposal API, but the debate is ongoing!
- We are still working on polishing up some UX issues.
- There are missing widgets and features. Notably the "text input" widget is still being developed.

If you're looking to experiment with building tooling for Bevy, enable it and take `feathers` for a test flight!
Let us know what problems you run into, and feel free to contribute missing widgets and bugs upstream.

If you can't wait to get your hands on `bevy_ui` widgets for your game,
we recommend copying the Feathers code into your project and start hacking away at it!
Feathers can serve as a helpful base to understand how to build and theme widgets in Bevy UI. It also illustrates how to use our new "headless" widget set: `bevy_ui_widgets`.

## Realtime-filtered environment maps

{{ heading_metadata(authors=["@mate-h"] prs=[19076, 20529]) }}
(TODO: Embed screenshot of atmosphere-generated reflections)

> An environment map needs to be processed to be able to support uses beyond a simple skybox,

Reviewer: Way too confusing imo, even for me. I would focus more on user use cases, and how this partially replaces the hard-to-build gltf-ibl-sampler stuff + EnvironmentMapLight, and enables more dynamic use cases like having procedural atmospheres contribute lighting to the scene instead of just being a background. Mention specific components by name.

- It may be sensitive to rust/linker configuration. Dioxus is already pretty good about this though!

We have plans to further expand support, including making the upcoming [`bsn!` macro](https://github.com/bevyengine/bevy/pull/20158/) hot-reloadable (check out [this video](/news/bevys-fifth-birthday/#bevy-hot-reloading) of it in action!).

## Deep Learning Super Sampling (DLSS)

{{ heading_metadata(authors=["@JMS55", "@cart"] prs=[19864, 19817, 20565]) }}

For users with NVIDIA RTX GPUs, Bevy now offers yet another form of anti-aliasing: DLSS.

Try it out by running Bevy's anti_aliasing example: `cargo run --example anti_aliasing --features dlss --release` (after performing setup from <https://github.com/bevyengine/dlss_wgpu>).

Additionally, we've open sourced <https://github.com/bevyengine/dlss_wgpu> as a standalone crate to help other wgpu-based renderers integrate DLSS.

Compared to Bevy's built-in TAA, DLSS:

- Produces higher quality and more stable images
- Supports upscaling in addition to anti-aliasing, leading to much cheaper render times, particularly when used with GPU-heavy features like Bevy Solari
- Requires a NVIDIA RTX GPU

> - Currently requires running via the Vulkan backend on Windows/Linux (no macOS, web, or mobile support)

Reviewer: ```suggestion
- Currently requires running via the Vulkan backend on Windows/Linux (no DirectX, macOS, web, or mobile support)
```

- Enable Bevy's `dlss` feature
- Insert the `DlssProjectId` resource before `DefaultPlugins` when setting up your app
- Check for the presence of `Option<Res<DlssSuperResolutionSupported>>` at runtime to see if DLSS is supported on the current machine
- Add the `Dlss` component to your camera entity, optionally setting a specific `DlssPerfQualityMode` (defaults to `Auto`)
- Optionally add sharpening via `ContrastAdaptiveSharpening`
- Custom rendering code, including third party crates, should account for the optional `MainPassResolutionOverride` to work with DLSS (see the `custom_render_phase` example)

Note that DLSS integration is expected to have some bugs in this release related to certain rendering effects not respecting upscaling settings, and possible issues with transparencies or camera exposure. Please report any bugs encountered.

Other temporal upscalers like AMD's FidelityFX™ Super Resolution (FSR), Intel's Xe Super Sampling XeSS (XeSS), and Apple's MTLFXTemporalScaler are not integrated in this release. However they all use similar APIs, and would not be a challenge to integrate in future releases.

> Support for other swapchain-related features like frame interpolation/extrapolation, latency reduction, or dynamic resolution scaling are not currently planned, but support for DLSS Ray Reconstruction for use in Bevy Solari _is_ planned for a future release.

Reviewer: ```suggestion
Support for other swapchain-related features like frame interpolation/extrapolation, latency reduction, or dynamic resolution scaling are not currently planned.
```

Reviewer: I did end up implementing DLSS-RR this cycle.

Special thanks to @cwfitzgerald for helping with the [`wgpu`](https://github.com/gfx-rs/wgpu) backend interop APIs.

## Raymarched atmosphere and space views

{{ heading_metadata(authors=["@mate-h"] prs=[20766]) }}

(Insert screenshot of space shot including volumetric shadows)

> Bevy's atmosphere now supports a raymarched rendering path that unlocks accurate views from above the atmosphere. This means **Bevy 0.17** now has two atmosphere rendering modes to choose from:

Reviewer: ```suggestion
Bevy's procedural atmosphere now supports a raymarched rendering path that unlocks accurate views from above the atmosphere. This means **Bevy 0.17** now has two atmosphere rendering modes to choose from:
```

any unsupported platforms to be supported upstream, sometimes it might not be possible. For this reason, there is a different implementation of this feature that works on all platforms.
It comes with some caveats with regards to project structure and might increase compile time, so it is better used as a backup solution. The detailed instructions on how to use this feature
can be found in this [`example`]. Types can also still be manually registered using `app.register_type::<T>()`.

[`Reflect`]: https://docs.rs/bevy/0.17.0/bevy/prelude/trait.Reflect.html
[`inventory`]: https://github.com/dtolnay/inventory
[`example`]: https://github.com/bevyengine/bevy/tree/release-0.17.0/examples/reflection/auto_register_static
[`register_type`]: https://docs.rs/bevy/0.17.0/bevy/prelude/struct.App.html#method.register_type

## Virtual Geometry BVH culling

{{ heading_metadata(authors=["@SparkyPotato", "@atlv24"] prs=[19318]) }}

(TODO: Embed example screenshot here)

Bevy's virtual geometry has been greatly optimized with BVH-based culling, making the cost of rendering nearly independent of scene geometry.

> Comparing the sample scene shown above with 130k dragon instances to one with over 1 million instances, total GPU rendering time only increases by 30%.

Reviewer: This sentence is confusing, and kind of redundant when we have perf numbers lower down.

This also gets rid of the previous cluster limit that limited the world to 2^24 clusters (about 4 billion triangles).
There are now _no_ hardcoded limits to scene size. In practice you will only be limited by asset VRAM usage (since streaming is not yet implemented),

> and total instance count due the current code requiring all instances to be re-uploaded to the GPU every frame.

Reviewer: ```suggestion
and total instance count (again due to temporary limitations we're working on improving).
```

Comparing GPU times to **Bevy 0.16** on a much smaller scene with 1,300 instances, previously the full render took 2.2 ms, whereas now in **Bevy 0.17** it takes 1.3 ms.

## Frame Time Graph

{{ heading_metadata(authors=["@IceSentry", "@Zeophlite"] prs=[12561, 19277]) }}

(TODO: Embed frame time graph gif from 12561)

When measuring a game's performance, just seeing a number is often not enough. Seeing a graph that shows the history makes it easier to reason about performance. **Bevy 0.17** introduces a new visual "frame time graph" to solve this problem!

To display the frame time graph, enable the `bevy_dev_tools` cargo feature and add in `FpsOverlayPlugin`:

This displays "frame time" not "frames per second", so a longer frame time results in a bigger and wider bar. The color also scales with that frame time. Red is at or bellow the minimum target fps and green is at or above the target maximum frame rate.
Anything between those 2 values will be interpolated between green and red based on the frame time.

The algorithm is highly inspired by [Adam Sawicki's article on visualizing frame times](https://asawicki.info/news_1758_an_idea_for_visualization_of_frame_times).

> ## `Text2d` Shadows

Reviewer: Might be nice to group all the UI stuff together. Gradients are way further up.

which is counterintuitive and very annoying when working across different art pipelines.

To remedy this, users can now change the import behavior to instead favor correct `Transform::forward()` semantics for models.
The downside is that glTF cameras and lights that have a global identity transform in glTF will now look to +Z instead of -Z in Bevy.
This should not be a problem in many cases, as the whole scene is rotated so that the end result on your screen will be rendered the exact same way.

To globally opt into the behavior that favors glTF models over glTF cameras, you can set `GltfPlugin::use_model_forward_direction`:

```rust
App::new()
    .add_plugins(DefaultPlugins.set(GltfPlugin {
        use_model_forward_direction: true,
        ..default()
    }))
    .run();
```

You can also control this on a per-asset-level:

```rust
let handle = asset_server.load_with_settings(
    "fox.gltf#Scene0",
    |settings: &mut GltfLoaderSettings| {
        settings.use_model_forward_direction = Some(true);
    },
);
```

Setting the above to `None` will fall back to the global setting taken from `GltfPlugin::use_model_forward_direction`.

> ## `Text2d` window independent scale factor support

Reviewer: I feel like we might want to cut this section.

{{ heading_metadata(authors=["@Icksonpe"] prs=[20656]) }}

In previous versions of bevy, text rendered using `Text2d` would always use the scale factor of the primary window regardless of its render target. In 0.17, `Text2d` glyphs are rasterized to match the scale factor of the render target where the text is to be drawn.

`Text2d` is still limited to generating only one text layout per `Text2d` entity. If a `Text2d` entity is simultaneously rendered to multiple targets with different scale factors then the maximum of the target scale factors is used.

> ## Consistent `*Systems` naming convention for system sets

Reviewer: I would move this all the way to the bottom.
